File: traffic/src/main/java/com/hsryuuu/traffic/coupon/repository/CouponRepository.java
package com.hsryuuu.traffic.coupon.repository;

import com.hsryuuu.traffic.coupon.entity.Coupon;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;

public interface CouponRepository extends JpaRepository<Coupon, Long> {

    boolean existsByCouponEventIdAndUserId(Long couponEventId, Long userId);

    List<Coupon> findAllByUserId(Long userId);
}

File: high-traffic/src/main/kotlin/com/hsryuuu/hightraffic/coupon/repository/CouponRepository.kt
package com.hsryuuu.hightraffic.coupon.repository

import com.hsryuuu.hightraffic.coupon.entity.Coupon
import io.lettuce.core.dynamic.annotation.Param
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query
import java.util.Optional

interface CouponRepository : JpaRepository<Coupon, Long>{

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT c FROM Coupon c WHERE c.id = :id")
    fun findByIdWithPessimisticLock(@Param("id") id: Long): Optional<Coupon>
}
